### DISCUSSION TITLE
Create playlist based on tag/metadata

### DISCUSSION THREAD

**Discussion by triangle-stream** [savonet/liquidsoap#1755] [Q&A]
Hello,

Is there any way for Liquidsoap to generate a playlist based on files' id3tag/metadata? If not, is there any resource that you can suggest in order to make this possilbe (scripts, docs, etc.)?

Thank you,
Marco.

**Comment by smimram** [comment]
One possibility is to use the `check_next` parameter of `playlist`. For instance,
```
def check(r)
  request.read_metadata(r)
  m = request.metadata(r)
  m["genre"] == "Rock"
end
s = playlist(check_next=check, "~/Music")
output(s)
```
another possibility is to use `playlist.list` and generate the list of file from `playlist.files`, for instance
```
def check(f)
  m = file.metadata(f)
  if m["genre"] == "Dance" then
    print("Keeping #{f}.")
    true
  else
    false
  end
end
l = playlist.files("~/Music/misc")
l = list.filter(check, l)
s = playlist.list(l)
output(s)
```

**Comment by triangle-stream** [comment]
Really thank you @smimram!

May I ask you a little help with this? I'm editing an existing .liq config used by Azuracast. I only have the ability to edit this in order to add some functionalities to something they call "Advanced Playlist" in which you can customize parts of the .liq.

Said that, here is how I tried to add the check_next based on your answer (unfortunately noy working):
```
def check(r)
  request.read_metadata(r)
  m = request.metadata(r)
  m["genre"] == "triste"
end
playlist_test= playlist(id="playlist_test",check_next=check,mode="randomize",reload_mode="watch",conservative=true,default_duration=10.,length=20.,"/var/azuracast/stations/alfredo_-_copy/playlists/playlist_test.m3u")
```

that is just before these lines that are not directly editable:
```
playlist_test = playlist(id="playlist_test",mode="randomize",reload_mode="watch",conservative=true,default_duration=10.,length=20.,"/var/azuracast/stations/alfredo_-_copy/playlists/playlist_test.m3u")
playlist_test = audio_to_stereo(id="stereo_playlist_test", playlist_test)
playlist_test = cue_cut(id="cue_playlist_test", playlist_test)
ignore(playlist_test)
```

I know that i'm wrong in something, could you help me in identifying at which point?

**Comment by smimram** [comment]
What do you mean exactly by "not working"? Is there a particular error message? Which version of Liquidosap are you using?

**Comment by triangle-stream** [comment]
I'm sorry. I found this error in liquidsoap log:
`Error 4: Undefined variable request.read_metadata`

Liquidsoap version is 1.4.4

For not working i mean that no playlist is written into the file playlist_test.m3u specified in config.

[Here](https://pastebin.com/kXwe4bvR) is the full config.

**Comment by smimram** [comment]
Can you try without the `read_metadata` line? Something like
```
def check(r)
  m = request.metadata(r)
  log("Checking request:\n#{m}\n\n")
  m["genre"] == "Rock"
end
s = playlist(check_next=check, "~/Music")
output(s)
```
It might work in 1.4.4.

I also added a printing of recieved metadata to help debugging.

**Comment by triangle-stream** [comment]
I can confirm that error 4 has gone away. 

Now we have a warn:
`Warning 4: Unused variable playlist_test`

I also fixed a typo on line 34 of config (no space between var and "=").

Am i wrong in thinking that liquidsoap read its config from top to bottom? In this case, is it possible that these following statements redefine the playlist?
```
playlist_test = playlist(id="playlist_test",mode="randomize",reload_mode="watch",conservative=true,default_duration=10.,length=20.,"/var/azuracast/stations/alfredo_-_copy/playlists/playlist_test.m3u")
playlist_test = audio_to_stereo(id="stereo_playlist_test", playlist_test)
playlist_test = cue_cut(id="cue_playlist_test", playlist_test)
ignore(playlist_test)
```

No log lines in logfile about the "Checking request:..." but maybe this has to do with this line:
`set("log.file", false)`

**Comment by triangle-stream** [comment]
Ok my fault: i used that playlist var, now I have logs and it is working! 

Thank you @smimram, I'll write here as i will have something good that other could also use.

**Comment by triangle-stream** [comment]
Sorry for so much posts: using the request.metadata, confirming that the playlist now is playing tracks on this criteria, is it possible to write that playlist into a specified file? I've also noticed that i have to restart liquidsoap in order to refresh the playlist but, maybe, this could be avoided by including the watch directive.

Thank you.
